walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,107 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Iterator;
+
+    // first iterate through and combine events that overlap
+    // subtract combined events calendar from whole day
+    // remove time ranges that are too short 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+        List<TimeRange> availableRanges = new ArrayList<>();
+        availableRanges.add(TimeRange.WHOLE_DAY);
+
+        Collection<TimeRange> bookedRanges = combinedMandatoryRanges(events, request);
+
+        /* The bookedRanges is already sorted, so iterating from the earliest bookedRange, 
+            each subsequent bookedRange will have a conflict with the latest avaliable range, as we start with the whole day */
+        for (TimeRange booked : bookedRanges) {
+            TimeRange lastInavailable = availableRanges.get(availableRanges.size()-1);
+            // if (lastInavailable.overlaps(booked)){ //if statement unneeded
```

Comment:
If it's unneeded, it should be removed rather than commented out

Reply:
Okay.